### PR TITLE
Reduce text layout CPU usage when DWrite analysis is not needed

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

It takes many CPU cycles to calculate text layout & font fallback even it's not really necessary. This may seem a little aggresive but it indeed dramatically boost performance.

## References

#806 

## PR Checklist
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan.

## Detailed Description of the Pull Request / Additional comments

With the help of #2937 and this PR, vim performance is way more smooth than before.

### CODE REVIEW

**Comment by skyline75489** [comment]
Seems something wrong with the format. But I can't really tell what it is.

**Comment by zadjii-msft** [comment]
@skyline75489 It's _probably_ the `all_of` call, if I'm spitballing. Running `tools\runformat` in the root of the repo should fix it for you though :)

**Comment by skyline75489** [comment]
@zadjii-msft Thanks! I'll give it a try. 

This PR is crucial almost the entire "make cmatrix great again" movement. Would love to see it merged.

File: src/renderer/dx/CustomTextLayout.cpp [hunk 1]
```diff
@@ -129,13 +129,30 @@ CustomTextLayout::CustomTextLayout(gsl::not_null<IDWriteFactory1*> const factory
...
 
-        // Call each of the analyzers in sequence, recording their results.
-        RETURN_IF_FAILED(_analyzer->AnalyzeLineBreakpoints(this, 0, textLength, this));
-        RETURN_IF_FAILED(_analyzer->AnalyzeBidi(this, 0, textLength, this));
-        RETURN_IF_FAILED(_analyzer->AnalyzeScript(this, 0, textLength, this));
-        RETURN_IF_FAILED(_analyzer->AnalyzeNumberSubstitution(this, 0, textLength, this));
-        // Perform our custom font fallback analyzer that mimics the pattern of the real analyzers.
-        RETURN_IF_FAILED(_AnalyzeFontFallback(this, 0, textLength));
+        const bool whiteSpacesOnly = std::all_of(
+            _text.begin(), _text.end(), [](wchar_t c) noexcept {
```
Comment (zadjii-msft) [inline src/renderer/dx/CustomTextLayout.cpp]: Out of curiosity, can we just pass `&iswspace` as the function? (my c++17-fu is not strong so this is mostly an educational question).

(and the same below)

**Comment by miniksa** [comment]
1. For the whitespace part:
- Using the CRT function iswspace is uncomfortable to me because it is doing a locale detection in a way that is completely different and potentially out of sync from whatever locale settings we're using in the rest of DirectWrite.
- We honestly shouldn't be dispatching draw requests for whitespace at all. This is the fault of https://github.com/microsoft/terminal/blob/b9233c03d12bc7addab16647dba07196bd554f6f/src/renderer/dx/DxRenderer.cpp#L734 which I put in there to always force full frame drawing until we could just get things to work. #778 to restore differential drawing would remove that line and eliminate a grand majority of whitespace drawing through the existing dirty region tracking.

2. For the ascii only part:
- Again, not a fan of using the CRT functions.
- If this method's really expensive, we can probably avoid it another way. I'm pretty sure `_analyzer` has a `GetTextComplexity()` method that would probably indicate to us whether or not we need to go into the whole expensive calculations or not. For that matter, the complexity function might be able to gate all of these calls.

I would prefer to always lean on further portions of the DirectWrite framework to analyze the text, determine complexity, or figure out which regions we can discard before calling an expensive function. I really don't want to mix in other technologies like the CRT here. If we don't know the right way to do something in DWrite, I'd rather lean on our internal contacts on those teams to help us figure it out than try to reverse it with a mix of other technologies.

Overall, though, I think this particular CPU concern is just majorly an artifact of #778 being incomplete.

**Comment by skyline75489** [comment]
@miniksa I agree with you on that #778 is the ultimate solution for overall performance. However, I'd argue that this optimization is still needed even #778 is complete. There are still many situations that requires entire screen redrawing(e.g. vim, cat, and less). Differential drawing won't optimize away the CPU burden with all those DWrite analysis.

I'm not a DWrite expert and I chose CRT on this one. Seems `GetTextComplexity()` also works great on this one.

**Comment by skyline75489** [comment]
The concern I have on this, is what will `GetTextComplexity` behave under more sophiscated condition. For example, different locale and font with ligatures support. If we can trust `GetTextComplexity` on this, I think it's safe to use it for a short cut.

**Comment by skyline75489** [comment]
I found a catch. If I open `cmd.exe` the Chinese character is not correctly rendered.

**Comment by skyline75489** [comment]
I added the check of text read length to eliminate the issue with `cmd.exe`.

**Comment by madig** [comment]
Hrm, I remember someone saying somewhere that Safari uses a similar trick to save on layouting and that it can lead to tricky edge-cases with font features. A terminal is probably less likely to deal with OpenType Layout data in fonts, so it might be okay 🤔

**Pull request merged**